### PR TITLE
Added SRTO_BINDTODEVICE option

### LINKED ISSUE #558: [FR] Add SO_BINDTODEVICE UDP socket option
Why it is not possible to use the ADAPTER setting in the CALLER mode? The ADAPTER setting is working for LISTENER and RENDEVOUS, but I would like to use the ADAPTER setting also in the CALLER mode to specify an interface for the outgoing datagrams. So at the moment I have to use the RENDEVOUS mode instead of the CALLER mode, but why?

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #558

### CODE REVIEW

File: srtcore/channel.cpp [hunk 1]
```diff
@@ -252,6 +252,21 @@ void CChannel::setUDPSockOpt()
       }
 #endif
 
+#ifdef SRT_ENABLE_BINDTODEVICE
+      if (m_BindToDevice != "")
```
Comment (maxsharabayko) [inline srtcore/channel.cpp:256]:
```suggestion
      if (!m_BindToDevice.empty())
```

File: srtcore/queue.h [hunk 2]
```diff
@@ -549,6 +549,7 @@ struct CMultiplexer
 #ifdef SRT_ENABLE_IPOPTS
    int m_iIpTTL;
    int m_iIpToS;
+   std::string m_BindToDevice;
```
Comment (maxsharabayko) [inline srtcore/queue.h:552]: `#ifdef SRT_ENABLE_BINDTODEVICE` is missing.
Is it used?

File: srtcore/channel.cpp [hunk 3]
```diff
@@ -252,6 +252,21 @@ void CChannel::setUDPSockOpt()
...
+      {
+          if (m_BindAddr.family() != AF_INET)
+          {
+              LOGC(mglog.Error, log << "SRTO_BINDTODEVICE can only be set with AF_INET connections");
+              throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+          }
+
+          if (0 != ::setsockopt(m_iSocket, SOL_SOCKET, SO_BINDTODEVICE, m_BindToDevice.c_str(), m_BindToDevice.size()))
```
Comment (maxsharabayko) [inline srtcore/channel.cpp:264]: Joda style? :)

Comment (ethouris) [reply srtcore/channel.cpp:264]: This way it's clearer at the beginning that the condition is for failure.

**Comment by maxsharabayko** [comment]
- [ ] Retrieve the "bind to device" value from a socket (get socket option).

**Comment by ethouris** [comment]
As per necessity of this field in `CMultiplexer` class, see api.cpp:2445.

File: srtcore/queue.h (hunk 2 above)
Comment (ethouris) [reply srtcore/queue.h:549]: Fixed.

**Comment by ethouris** [comment]
`SRT_ENABLE_BINDTODEVICE` is not a build option. This part of the code is enabled always on Linux and not on other platforms, automatically.

File: docs/API.md [hunk 4]
```diff
@@ -487,6 +487,25 @@ This option list is sorted alphabetically. Note that some options can be
 either only a retrieved (GET) or specified (SET) value.
 
 
+| OptName               | Since | Binding | Type     | Units  | Default  | Range  |
+| --------------------- | ----- | ------- | -------- | ------ | -------- | ------ |
+| `SRTO_BINDTODEVICE`   | 1.5.0 | pre     | `string` |        |          |        |
+
+- Refets to `SO_BINDTODEVICE` system socket option for `SOL_SOCKET` level. This
```
Comment (stevomatthews) [inline docs/API.md:494]:
```suggestion
- Refers to the `SO_BINDTODEVICE` system socket option for `SOL_SOCKET` level. This
```

File: srtcore/core.cpp [hunk 5]
```diff
@@ -1159,6 +1186,27 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
...
+#ifdef SRT_ENABLE_BINDTODEVICE
+        if (optlen < IFNAMSIZ)
+            throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+
+        if (m_bOpened)
+            if (m_pSndQueue->getBind(((char*)optval), optlen))
+            {
+                optlen = strlen((char*)optval);
+                break;
+            }
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1199]:
```suggestion
            if (m_bOpened && m_pSndQueue->getBind(((char*)optval), optlen))
            {
                optlen = strlen((char*)optval);
                break;
            }
```

File: srtcore/core.cpp [hunk 6]
```diff
@@ -1159,6 +1186,26 @@ void CUDT::getOpt(SRT_SOCKOPT optName, void *optval, int &optlen)
...
+            throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);
+
+            if (m_bOpened && m_pSndQueue->getBind(((char*)optval), optlen))
+            {
+                optlen = strlen((char*)optval);
+                break;
+            }
+
+        // Fallback: return from internal data
+        strcpy(((char*)optval), m_BindToDevice.c_str());
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1201]: Alignment.

```suggestion
        if (optlen < IFNAMSIZ)
            throw CUDTException(MJ_NOTSUP, MN_INVAL, 0);

        if (m_bOpened && m_pSndQueue->getBind(((char*)optval), optlen))
        {
            optlen = strlen((char*)optval);
            break;
        }

        // Fallback: return from internal data
        strcpy(((char*)optval), m_BindToDevice.c_str());
```

**Pull request merged**